pkg/kv/kvserver/rangefeed/buffered_registration.go

```diff
@@ -120,6 +120,10 @@ func (br *bufferedRegistration) publish(
 	if br.mu.overflowed {
 		return
 	}
+	if br.mu.disconnected {
```

Comment:
in the commit message, what do you men by "adding it to the buffer is only more work later"? What exactly happens to an event in a disconnected buffer?

Reply:
> What exactly happens to an event in a disconnected buffer?

After `registration.disconnect()` is called, additional events may still be buffered from the processor https://github.com/cockroachdb/cockroach/blob/a4cbf7d7ec37fbb351e453bf2c7314bac9cbfe9a/pkg/kv/kvserver/rangefeed/buffered_registration.go#L125. The buffer is not drained until `reg.Unregister` is called which officially unsubscribe the registration from the processor and triggers the draining process https://github.com/cockroachdb/cockroach/blob/a4cbf7d7ec37fbb351e453bf2c7314bac9cbfe9a/pkg/kv/kvserver/rangefeed/registry.go#L334-L341. `reg.Unregister` currently happens in the separate per-registration goroutine after `runOutputLoop` returns here https://github.com/cockroachdb/cockroach/blob/3bf870a87e8dd943fe9407dce277e20ae1718d6c/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L351-L357. They should all occur within a short time frame.

In this PR, we changed the flow a bit. After `registration.disconnect()` is called, no additional events can be buffered. Buffer draining now happens within the defer statement of `runOutputLoop`, which may occur before the processor's official unregistration (`reg.Unregister`).


> in the commit message, what do you men by "adding it to the buffer is only more work later"?

I believe this is what Steven meant by `adding it to the buffer only creates more work later.  For events buffered after `r.Disconnect()` but before `p.reg.Unregister()` happens, they are unlikely to be published, as `runOutputLoop` would be in the process of being torn down. Therefore, these events will just be buffered only to be drained later during `p.reg.Unregister()`. Instead, we can simply avoid buffering these events during by checking for r.disconnected.

Aside - currently, `p.reg.Unregister()` runs in a separate per-registration goroutine, but we plan to make it async and delegate the task to the processor goroutine here https://github.com/cockroachdb/cockroach/blob/9a1fe4e3f8758a6cc5c8fb47273dfd3d7865fd51/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L167. Moving buffer draining to the end of separate goroutine `runOutputLoop` would also help reduce processor's work.
<!-- Reviewable comment -OC9SlOI91S9c0crqTj7:bb5ktpe -->
<!-- Sent from Reviewable.io -->